Added a new ProcessorTopologyTestDriver class that makes it easier to unit test a ProcessorTopology

rhauch opened the pull request:
The new test driver class uses a supplied `TopologyBuilder` and a `StreamingConfig` to set up a topology using the existing `StreamTask`, which is the same component used in the Kafka Streams framework for each thread. However, rather than rely upon a real Kafka broker, the test driver simulates a broker with mock consumers and producers. This required a few changes to `ProcessorContextImpl`, `StreamTask`, and `StreamThread` so that the latter can pass in the actual consumer used to restore state; this allows the test driver to pass a mock consumer into `StreamTask`. Without this change, the test driver could not run topologies that used state stores.

(Updated version of #37)

guozhangwang commented:
@rhauch Could you take a look at the current ProcessorTopologyTest and see if it is necessary to rebase this change on top of the streaming branch?

rhauch commented:
@guozhangwang, looking now.

rhauch commented:
@guozhangwang, the first of my 2 original commits was superseded by the latest updates on the `streaming` branch. So I rebased on latest from `streaming`, removed the first commit that fixed the unsubscribe `ProcessorStateManager`, ran all tests on `streaming` (pass for me locally), and updated this PR.

rhauch commented:
@guozhangwang, I rebased the PR again, and it is again up-to-date with `streaming`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -224,9 +224,6 @@ public void close(Map<TopicPartition, Long> ackedOffsets) throws IOException {
             checkpoint.write(checkpointOffsets);
         }
 
-        // close the restore consumer
```
Is this intentional?

rhauch replied:
Closing the consumer used for restoring state was moved to [here](https://github.com/confluentinc/kafka/pull/57/files#diff-045aeaddb4232a85a8560186b4901e69R207), which is in the same class that now creates the consumer.

rhauch commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -224,9 +224,6 @@ public void close(Map<TopicPartition, Long> ackedOffsets) throws IOException {
 
-        // close the restore consumer
-        restoreConsumer.close();
-
```
Closing the consumer used for restoring state was moved to [here](https://github.com/confluentinc/kafka/pull/57/files#diff-045aeaddb4232a85a8560186b4901e69R207), which is in the same class that now creates the consumer.

rhauch commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -109,11 +109,6 @@ public StreamTask(int id,
-                config.getConsumerConfigs(),
-                new ByteArrayDeserializer(),
-                new ByteArrayDeserializer());
-
```
For testability purposes, its important to be able to pass in `MockConsumer` and  `MockProducer` instances rather than hard-code the creation of them.

ymatsuda replied:
I agree. :+1:

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -70,14 +69,15 @@
      * @param topology              the instance of {@link ProcessorTopology}
      * @param config                the {@link StreamingConfig} specified by the user
      */
-    @SuppressWarnings("unchecked")
```
Some of the latest commits are reverted.

rhauch replied:
@guozhangwang, I'm not sure what you mean. The base of this PR is 25a869b, which appears to be the latest commit on `streaming`. I can get rid of this PR's removal of line 77 if need be.

guozhangwang commented on streams/src/test/java/org/apache/kafka/test/ProcessorTopologyTestDriver.java:
```diff
@@ -0,0 +1,316 @@
+     * @param storeNames the names of the stores that this
+     * @return the mock consumer; never null
+     */
+    protected MockConsumer<byte[], byte[]> createRestoreConsumer(int id, String... storeNames) {
```
Where is this MockConsumer used? If it is not used anywhere could we just let the StreamTask create its own restore consumer?

rhauch replied:
It's used in the constructor on [line 146](https://github.com/confluentinc/kafka/pull/57/files#diff-a2bf902db43bffc65fe93b819acc218aR146). It's protected in case the subclass wants to override.

guozhangwang commented on streams/src/test/java/org/apache/kafka/test/ProcessorTopologyTestDriverTest.java:
```diff
@@ -0,0 +1,323 @@
+    /**
+     * Abstract base Processor.
+     */
+    private static abstract class FauxProcessor implements Processor<String, String> {
```
This is the same as KStreamProcessor.

rhauch replied:
Yes, but `KStreamProcessor` is package protected. Shall I can make it public and use it instead?

ymatsuda replied:
I think it is fine to have FauxProcessor here.

guozhangwang commented:
Thanks for the patch. I was originally thinking it reverts some commits since it moves the restore consumer construction out of stream task again. Since the restore consumer is not yet used, I suggest we move it back to StreamTask.

Another general comment is to move the tests in ProcessorTopologyTestDriverTest to ProcessorTopologyTest, and move the simple processors into the ProcessorTopologyTestDriver.

rhauch commented:
Yes, it is possible to move the restore consumer creation into `StreamTask`, but this limits what can be tested without a real Kafka broker running.

The restore consumer is used by the `ProcessorStateManager` when the `ProcessorContextImpl.register(StateStore store, RestoreFunc restoreFunc)` method is called by the `MeteredKeyValueStore` constructor. So when testing a topology with processors that use a `KeyValueStore`, the restore consumer (if a real one is used) attempts to connect to a real Kafka broker. IOW, moving the restore consumer construction out of StreamTask is necessary so that we can test all varieties of topologies. Does that make sense?

> Another general comment is to move the tests in ProcessorTopologyTestDriverTest to ProcessorTopologyTest, and move the simple processors into the ProcessorTopologyTestDriver.

Are you suggesting that `ProcessorTopologyTest` should use `ProcessorTopologyTestDrivr`? If so, then I agree that will clean up the tests.

ymatsuda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -97,12 +98,18 @@ public void testPartitionAssignmentChange() throws Exception {
         builder.addSource("source2", "topic2");
 
-        StreamThread thread = new StreamThread(builder, config, producer, consumer, new SystemTime());
+        StreamThread thread = new StreamThread(builder, config, producer, consumer, new SystemTime()) {
```
missing restoreConsumer in the arguments?

rhauch replied:
Ack.

rhauch commented:
@guozhangwang, @ymatsuda: I added a commit that fixes the `StreamThreadTest` and moves the `ProcessorTopologyTestDriver` tests into the `ProcessorTopologyTest` class as suggested. I left the creation of the restore consumer in its new home in `StreamThread` so that all of the `ProcessorTopologyTest` methods pass. Running `./gradlew clean test` passes successfully.

Please holler if there's anything else I should change.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:
```diff
@@ -53,4 +135,221 @@ public void testTopologyMetadata() {
+    }
+
+    /**
+     * A processor that simply forwards all messages to all children.
```
Maybe we can move the KStreamProcessor into the processor.internals package and merge it with this Processor? @ymatsuda @rhauch

rhauch replied:
@guozhangwang, I can do that if you prefer, but IMO moving `KStreamProcessor` outside of the package where it's primarily used unnecessarily complicates the KStream internals. Perhaps the fact that KStreams' processors, these test processors, and other `Processor` implementations of clients might want to manage the `ProcessorContext` instance in a similar way suggests there should be an public `AbstractProcessor` implementation in the `processor` package. Thoughts? I'd be happy to add that to this PR in another commit.

ymatsuda replied:
+1 for AbstractProcessor

guozhangwang replied:
lgtm.

rhauch commented:
@guozhangwang, @ymatsuda: I've added a commit that creates a public `org.apache.kafka.streams.processor.AbstractProcessor`, replacing `KStreamProcessor` (wherever it was used) and the `FauxProcessor` in the `ProcessorTopologyTest` class. The JavaDoc explains to subclass developers the intent of the methods and when its `context()` method will return a non-null context. The commit also adds JavaDoc for `Processor`.

If you don't like this last commit, I can revert it. (BTW, it would allow me to remove a similar class in my own project that uses the streams prototype branch.)

guozhangwang commented:
Thanks, LGTM!
